File: structural/decorator/request_pipeline/go/request_pipeline/request_pipeline_test.go
package request_pipeline

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestBaseHandler(t *testing.T) {
	request := Request{
		Headers: map[string]string{"Content-Type": "application/json"},
		Body:    map[string]interface{}{"data": "test"},
		Path:    "/api/test",
		Method:  "POST",
	}

	handler := &BaseHandler{}
	response := handler.Handle(request)

	assert.Equal(t, 200, response.StatusCode)
	assert.Equal(t, "Request processed successfully", response.Body["message"])
}

func TestAuthMiddleware(t *testing.T) {
	request := Request{
		Headers: map[string]string{"Content-Type": "application/json"},
		Body:    map[string]interface{}{"data": "test"},
		Path:    "/api/test",
		Method:  "POST",
	}

	// Test unauthorized request
	handler := NewAuthMiddleware(&BaseHandler{})
	response := handler.Handle(request)

	assert.Equal(t, 401, response.StatusCode)
	assert.Equal(t, "Unauthorized", response.Body["error"])

	// Test authorized request
	request.Headers["Authorization"] = "Bearer valid-token"
	response = handler.Handle(request)

	assert.Equal(t, 200, response.StatusCode)
	assert.Equal(t, "Request processed successfully", response.Body["message"])
}

func TestRateLimitMiddleware(t *testing.T) {
	request := Request{
		Headers: map[string]string{"Content-Type": "application/json"},
		Body:    map[string]interface{}{"data": "test"},
		Path:    "/api/test",
		Method:  "POST",
	}

	handler := NewRateLimitMiddleware(&BaseHandler{}, 2)

	// First request should succeed
	response := handler.Handle(request)
	assert.Equal(t, 200, response.StatusCode)

	// Second request should succeed
	response = handler.Handle(request)
	assert.Equal(t, 200, response.StatusCode)

	// Third request should fail
	response = handler.Handle(request)
	assert.Equal(t, 429, response.StatusCode)
	assert.Equal(t, "Too many requests", response.Body["error"])
}

func TestMiddlewareChain(t *testing.T) {
	request := Request{
		Headers: map[string]string{"Content-Type": "application/json"},
		Body:    map[string]interface{}{"data": "test"},
		Path:    "/api/test",
		Method:  "POST",
	}

	// Create a chain: Auth -> RateLimit -> BaseHandler
	handler := NewAuthMiddleware(
		NewRateLimitMiddleware(
			&BaseHandler{},
			2,
		),
	)

	// Test with valid auth
	request.Headers["Authorization"] = "Bearer valid-token"
	response := handler.Handle(request)
	assert.Equal(t, 200, response.StatusCode)

	// Test with invalid auth
	delete(request.Headers, "Authorization")
	response = handler.Handle(request)
	assert.Equal(t, 401, response.StatusCode)
} 

File: structural/decorator/request_pipeline/typescript/request_pipeline.test.ts
import {
  Request,
  Response,
  BaseHandler,
  AuthMiddleware,
  LoggingMiddleware,
  RateLimitMiddleware,
} from "./request_pipeline";

describe("Request Processing Pipeline", () => {
  let request: Request;

  beforeEach(() => {
    request = {
      headers: { "Content-Type": "application/json" },
      body: { data: "test" },
      path: "/api/test",
      method: "POST",
    };
  });

  test("BaseHandler processes request correctly", () => {
    const handler = new BaseHandler();
    const response = handler.handle(request);

    expect(response.statusCode).toBe(200);
    expect(response.body).toEqual({
      message: "Request processed successfully",
    });
  });

  test("AuthMiddleware blocks unauthorized requests", () => {
    const handler = new AuthMiddleware(new BaseHandler());
    const response = handler.handle(request);

    expect(response.statusCode).toBe(401);
    expect(response.body).toEqual({ error: "Unauthorized" });
  });

  test("AuthMiddleware allows authorized requests", () => {
    request.headers["Authorization"] = "Bearer valid-token";
    const handler = new AuthMiddleware(new BaseHandler());
    const response = handler.handle(request);

    expect(response.statusCode).toBe(200);
    expect(response.body).toEqual({
      message: "Request processed successfully",
    });
  });

  test("RateLimitMiddleware limits requests", () => {
    const handler = new RateLimitMiddleware(new BaseHandler(), 2);

    // First request should succeed
    let response = handler.handle(request);
    expect(response.statusCode).toBe(200);

    // Second request should succeed
    response = handler.handle(request);
    expect(response.statusCode).toBe(200);

    // Third request should fail
    response = handler.handle(request);
    expect(response.statusCode).toBe(429);
    expect(response.body).toEqual({ error: "Too many requests" });
  });

  test("Middleware chain works correctly", () => {
    // Create a chain: Auth -> RateLimit -> BaseHandler
    const handler = new AuthMiddleware(
      new RateLimitMiddleware(new BaseHandler(), 2)
    );

    // Test with valid auth
    request.headers["Authorization"] = "Bearer valid-token";
    let response = handler.handle(request);
    expect(response.statusCode).toBe(200);

    // Test with invalid auth
    delete request.headers["Authorization"];
    response = handler.handle(request);
    expect(response.statusCode).toBe(401);
  });
});
